doc/modules/clustering.rst

```diff
@@ -1300,3 +1300,244 @@ Drawbacks
+
+.. _unsupervised.silhouette_coefficient:
+
+Silhouette Coefficient
+----------------------
+
+Presentation and usage
+~~~~~~~~~~~~~~~~~~~~~~
+
+.. _stability:
+
+Stability
+---------
+
+Presentation
+~~~~~~~~~~~~
+
+A number of clusters is relevant if the clustering algorithm finds similar
+results with small perturbations of the data. In this implementation, we use
+the clustering algorithm on 2 large overlapping subset of the data. If the
+number of clusters is relevant, data is both subset should be clustered in
+a similar way. Let :math:`E` be the initial data, :math:`E_1` and :math:`E_2`
+be subsets of :math:`E`, each containing :math:`80\%` of :math:`E`.
+Let :math:`C_1` and :math:`C_2` be clusters obtained on datasets :math:`E_1`
+and :math:`E_2` Let :math:`A_i` be the adjacency matrix:
+
+.. math:: \left(A_i\right)_{ij} = \left\{
+	      \begin{tabular}
+	      1 & \mbox{if i and j belong to the same cluster} \\ 
+	      0 & \mbox{otherwise}
+	      \end{tabular}
+	  \right. 
+
+On points belonging to :math:`E_1` and :math:`E_2`, the adjacency matrix
+should stay the same. We compute the similarity between these two clustering
+
+.. math:: s(C_1, C_2) = \frac{
+   \sum_{(x_i, x_j) \in E_1 \cap E_2}\left(A_1\right)_{ij}\left(A_2\right)_{ij}}
+   {\sqrt{\sum_{(x_i, x_j) \in E_1 \cap E_2}\left(\left(A_1\right)_{ij}\right)^2}
+   \sqrt{\sum_{(x_i, x_j) \in E_1 \cap E_2}\left(\left(A_2\right)_{ij}\right)^2}}
+   
+For all number of clusters :math:`k=2\dots k_{max}`, we perform :math:`N_{draws}` times:
+
+- Select 2 subsets :math:`E_1` and :math:`E_2`
+
+- Compute clustering :math:`C_1` and :math:`C_2`
+
+- Compute similarity :math:`s(C_1, C_2)`
+
+The chosen number of clusters is the one that has maximum average similarity
+
+Usage
+~~~~~
+
+Given a dataset and a clustering algorithm (a :class:`ClusterMixin`),
+stability returns the most stable number of clusters
+
+  >>> from sklearn.datasets import make_blobs
+  >>> from sklearn.cluster import KMeans
+  >>> from sklearn.metrics.cluster.stability import stability
+  >>> data, labels = make_blobs(n_samples=1000, centers=2, random_state=0)
+  >>> kmeans_model = KMeans()
+  >>> stability(data, kmeans_model, k_max=10, verbose=False)
+  2
+
+Advantages
+~~~~~~~~~~
+
+- Finds a number of clusters that is truly relevant to your data and your
+  clustering algorithm
+
+- The stability score, going from 0 to 1, can measure how well your data is clustered
+  in k groups
+
+
+Drawbacks
+~~~~~~~~~
+
+- Computational time
+
+.. topic:: References
+
+ * `"A stability based method for discovering structure in clustered data"
+   <http://www.researchgate.net/profile/Asa_Ben-Hur/publication/11435997_A_stability_based_method_for_discovering_structure_in_clustered_data/links/00b4953988b6d0233d000000.pdf>`_
+   Ben-Hur, A., Elisseeff, A., & Guyon, I., *Pacific symposium on biocomputing* (2001, December)
+
+.. _gap_statistic:
+
+Gap statistic
+-------------
+
+Presentation
```

Comment:
I don't like these "presentation" headers and I think I might have removed them in master. (I know you went by what was there).